### PR TITLE
Added SQL Server Receiver Metrics.

### PR DESCRIPTION
**Pull request by StefanKurek**
Adds SQL Server metrics receiver to ops agent.

### CODE REVIEW

**Comment by qingling128** [comment]
The public SQL server images are documented here: https://cloud.google.com/compute/docs/images/os-details#sql_server.

It looks like our test framework should [already handles sql images](https://github.com/GoogleCloudPlatform/ops-agent/blob/2c2538ab50b5766f890066a95329f2aa7d2244a1/integration_test/gce/gce_testing.go#L351).

Reading https://github.com/GoogleCloudPlatform/ops-agent/blob/2b52f26b5b5fc5411f442509c5965f2b9af9b7ae/integration_test/third_party_apps_test.go#L115, it might already trigger MSSQL tests on SQL server images today. (CC @martijnvans in case I get it wrong)

**Comment by StefanKurek** [comment]
> The public SQL server images are documented here: https://cloud.google.com/compute/docs/images/os-details#sql_server.
> 
> It looks like our test framework should [already handles sql images](https://github.com/GoogleCloudPlatform/ops-agent/blob/2c2538ab50b5766f890066a95329f2aa7d2244a1/integration_test/gce/gce_testing.go#L351).
> 
> Reading
> 
> https://github.com/GoogleCloudPlatform/ops-agent/blob/2b52f26b5b5fc5411f442509c5965f2b9af9b7ae/integration_test/third_party_apps_test.go#L115
> 
> , it might already trigger MSSQL tests on SQL server images today. (CC @martijnvans in case I get it wrong)

@qingling128 gotcha. So if tests are already running on sql server images, then all I'd have to do is add another line to remove "sqlserver" from the apps as well (when it's a windows machine but not a sql server one)

**Comment by StefanKurek** [comment]
@qingling128 so I'm not sure that I actually have the ability to add new sql images to test against (I may be mistaken but I think this is set in the PLATFORMS environment variable), but given that someone adds to this, I think this code should be ready for it.

**Comment by martijnvans** [comment]
i gave it a shot but I did it wrong, so I need to wait another hour for results. Will update here once the latest run has finished.

**Comment by martijnvans** [comment]
OK I sent a change to add sql-std-2019-win-2019 to our tested images list. That is making its way through our presubmit tests now.

**Comment by martijnvans** [comment]
Oh and by the way, we sould include https://github.com/GoogleCloudPlatform/ops-agent/pull/547 into this PR or merge it into this PR, your choice. If we don't have that change, then our tests will try to run the `sqlserver` application on all versions of windows. [Which is bad, right? I might not understand the situation right.]

**Comment by martijnvans** [comment]
Tests passed, but it didn't run sqlserver. This is because sqlserver isn't in here in your PR:

https://github.com/GoogleCloudPlatform/ops-agent/blob/master/integration_test/third_party_apps_data/agent/windows/supported_applications.txt

**Comment by StefanKurek** [comment]
> Tests passed, but it didn't run sqlserver. This is because sqlserver isn't in here in your PR:
> 
> https://github.com/GoogleCloudPlatform/ops-agent/blob/master/integration_test/third_party_apps_data/agent/windows/supported_applications.txt

@martijnvans I'm pretty sure it's there https://github.com/GoogleCloudPlatform/ops-agent/blob/74f646aaa1f5c64af117ad09682d158ebd32dd82/integration_test/third_party_apps_data/agent/windows/supported_applications.txt

**Comment by martijnvans** [comment]
I'm sorry, the real reason it didn't run is because my pending change to actually run tests on sql server windows hasn't been accepted yet. I'll ping some people and see if we can get that in.

File: confgenerator/confgenerator.go
Comment (igorpeshansky) [inline confgenerator/confgenerator.go:89]: Why are we adding this across the board? Is there really no way to confine this change to the sqlserver receiver? Does this not break existing metrics?

File: confgenerator/testdata/invalid/linux/metrics-receiver_sqlserver_non_windows/golden_error
Comment (igorpeshansky) [inline confgenerator/testdata/invalid/linux/metrics-receiver_sqlserver_non_windows/golden_error:1]: Other tests of this kind are named `metrics-receiver_invalid_type_*` (e.g., `metrics-receiver_invalid_type_mssql`). For consistency, this should be named `metrics-receiver_invalid_type_sqlserver`.

I can see us renaming these wholesale to `metrics-receiver_sqlserver_unsupported` (`metrics-receiver_iis_unsupported`, `metrics-receiver_mssql_unsupported`, `logging-receiver_windows_event_log_unsupported`) to ensure consistent test prefix, but I'd strongly suggest doing that in a separate PR, and making this one consistent with the existing tests.

File: apps/sqlserver.go
Comment (igorpeshansky) [inline apps/sqlserver.go:41]: Do you need this local? Can you just inline the `map` below?

```suggestion
	return []otel.Pipeline{{
		Receiver: otel.Component{
			Type:   "sqlserver",
			Config: map[string]interface{}{
				"collection_interval": r.CollectionIntervalString(),
			},
		},
```

File: docs/sqlserver.md
Comment (igorpeshansky) [inline docs/sqlserver.md:6]: Typo: `perforamnce`

Comment (igorpeshansky) [inline docs/sqlserver.md:6]: `it is required for it` is a bit jarring. How about just `it is required to run`?

**Comment by martijnvans** [comment]
OK if you try again, it should run MS SQL server this time. I'll kick it off.

File: docs/sqlserver.md
Comment (igorpeshansky) [inline docs/sqlserver.md:3]: I'll let @qingling128 make the final determination here, but does it make sense to call out the differences between this receiver and `mssql`? Otherwise it's going to be very confusing to customers to have two receivers with seemingly the same functionality. If we do call out the difference, we should add some text to the mssql docs as well.

File: docs/sqlserver.md
Comment (igorpeshansky) [inline docs/sqlserver.md:42]: `Percent of pages` or `Fraction of pages`?

File: docs/sqlserver.md
Comment (igorpeshansky) [inline docs/sqlserver.md:38]: The descriptions of these metrics say `Number of …`, but they are actually rates. Do we need to mention something about that in the description?

Alternatively, should (can?) we ingest these as sums and let the users compute the rates at query time?

File: docs/sqlserver.md
Comment (igorpeshansky) [inline docs/sqlserver.md:48]: Shouldn't this be Sum (or Cumulative)? Ditto for the next metric…

File: docs/sqlserver.md
Comment (igorpeshansky) [inline docs/sqlserver.md:50]: `percent_used`?

File: integration_test/third_party_apps_data/applications/sqlserver/metadata.yaml
Comment (igorpeshansky) [inline integration_test/third_party_apps_data/applications/sqlserver/metadata.yaml:14]: This needs to be updated, doesn't it?

**Review by igorpeshansky — CHANGES_REQUESTED** [review]
Some of my comments about metric names and types should probably have been raised earlier in the scoping doc. If that's the case, please feel free to refer me to where that discussion happened in that doc.

File: confgenerator/confgenerator.go
Comment (StefanKurek) [reply confgenerator/confgenerator.go:89]: @igorpeshansky that's a good call out that I am aware of. I could minimize it a bit possibly by only including that filter if the sql server pipeline exists....but it still would apply to everything being exported in that circumstance because there is only a single google exporter for all apps (as far as I know).

Comment (StefanKurek) [reply confgenerator/confgenerator.go:89]: To add a little more to this, the work that was done to the google exporter (to be able to handle resource attributes) allows it to be configured with filters in order to translate the OTEL resource attributes to labels. Outside of this, nothing was added to account for the issues of OTEL receivers with resource attributes.

So unless I'm missing something, this is the best way I could think of dealing with the issue without coming up with a significant change to how the ops agent handles pipelines (which I know is not great because this change is on a global exporter which affects all pipelines...hopefully mitigated by good resource attribute naming).

**Comment by martijnvans** [comment]
Test failed, I see this error in [ops_agent_logs.txt](https://storage.cloud.google.com/ops-agents-public-buckets-test-logs/prod/stackdriver_agents/testing/consumer/third_party_apps/presubmit_github/windows/773/20220422-104541/logs/TestThirdPartyApps_sql-std-2019-win-2019_sqlserver/ops_agent_logs.txt)

```
4/22/2022 6:43:31 PM  1 Error            failed to generate config files: logging receiver with type "sqlserver" is    
                                         not supported. Supported logging receiver types: [apache_access,              
                                         apache_error, cassandra_debug, cassandra_gc, cassandra_system, couchdb,       
                                         elasticsearch_gc, elasticsearch_json, files, fluent_forward, hadoop,          
                                         hbase_system, iis_access, kafka, mongodb, mysql_error, mysql_general,         
                                         mysql_slow, nginx_access, nginx_error, postgresql_general, rabbitmq, redis,   
                                         solr_system, syslog, tcp, tomcat_access, tomcat_system, wildfly_system,       
                                         windows_event_log, zookeeper_general].                
```

File: docs/sqlserver.md
Comment (StefanKurek) [reply docs/sqlserver.md:38]: @igorpeshansky Unfortunately all of the sql server windows performance counters are gauges. The descriptions were mostly pulled verbatim from Microsoft's documentation for the counters (even though again they are gauges), but we could change the descriptions easily enough if need be.

File: docs/sqlserver.md
Comment (StefanKurek) [reply docs/sqlserver.md:50]: @qingling128 and @jefferbrecht as approvers of the scope doc, what are you opinions?

File: docs/sqlserver.md
Comment (jefferbrecht) [reply docs/sqlserver.md:3]: Similar to the [new `iis` metrics](https://github.com/GoogleCloudPlatform/ops-agent/pull/534/files#diff-98d6406bd3b368722a5ecde561e446b313751caddc7c593b5ae130467728cbcfR37), this should be under the existing `mssql` receiver with a `receiver_version: 2` field, rather than as a separate `sqlserver` receiver.

File: integration_test/third_party_apps_data/applications/sqlserver/metadata.yaml
Comment (jefferbrecht) [reply integration_test/third_party_apps_data/applications/sqlserver/metadata.yaml:14]: This should be 2.15.0 (2.14.0 is likely going out before this is merged).

File: docs/sqlserver.md
Comment (StefanKurek) [reply docs/sqlserver.md:3]: This was not quite called out in the scope doc "This will not replace the existing mssql metric functionality and will be able to be run alongside it." But I'll wait for @qingling128's feedback and if you're all in agreement then I'll move forward with this pattern.

File: integration_test/third_party_apps_data/applications/sqlserver/metric_name.txt
Comment (jefferbrecht) [inline integration_test/third_party_apps_data/applications/sqlserver/metric_name.txt:1]: This file isn't used anymore and can be deleted. (The `representative: true` metric under `expected_metrics` replaces the old behaviour.)

File: confgenerator/testdata/invalid/linux/metrics-receiver_sqlserver_non_windows/golden_error
Comment (StefanKurek) [reply confgenerator/testdata/invalid/linux/metrics-receiver_sqlserver_non_windows/golden_error]: fixed

File: apps/sqlserver.go
Comment (StefanKurek) [reply apps/sqlserver.go]: changed

File: docs/sqlserver.md
Comment (StefanKurek) [reply docs/sqlserver.md]: fixed

Comment (StefanKurek) [reply docs/sqlserver.md]: changed

File: docs/sqlserver.md
Comment (StefanKurek) [reply docs/sqlserver.md]: changed

File: docs/sqlserver.md
Comment (StefanKurek) [reply docs/sqlserver.md]: changed

File: integration_test/third_party_apps_data/applications/sqlserver/metadata.yaml
Comment (StefanKurek) [reply integration_test/third_party_apps_data/applications/sqlserver/metadata.yaml]: updated

File: integration_test/third_party_apps_data/applications/sqlserver/metric_name.txt
Comment (StefanKurek) [reply integration_test/third_party_apps_data/applications/sqlserver/metric_name.txt]: deleted

File: confgenerator/confgenerator.go
Comment (igorpeshansky) [reply confgenerator/confgenerator.go:89]: @StefanKurek I may be missing something about how metrics are modified, but it looks like these resource fields would be added indiscriminately. If not, a comment with a pointer to why this works would be really helpful.

@jefferbrecht I suspect we might just not be running the expected metrics tests with the `sqlserver` pipeline present. One way to be absolutely sure is to build a prototype package from this branch, install it, enable the `sqlserver` receiver, and see if it starts getting API failures due to extraneous labels. I can try to do that tomorrow unless someone beats me to it.

For the record, the substitution of `.` to `_` happens [here](https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/main/exporter/metric/metric.go#L641-L643).

Comment (StefanKurek) [reply confgenerator/confgenerator.go:89]: @igorpeshansky I believe this is the area of code you're looking for https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/836e6fe9832c9f19146dabab29a8b55d9511459f/exporter/collector/monitoredresource.go#L201

Comment (jefferbrecht) [reply confgenerator/confgenerator.go:89]: IIUC the labels are only added if the prefix matches the attribute key. It will be added indiscriminately in the sense that it's not receiver-specific, but effectively this is the only receiver whose attribute keys contain that prefix. There's some documentation [here](https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/3a2c57209765f4752bd140a007612a81f58489c1/exporter/collector/config.go#L92-L94), and the code for matching/adding is [here](https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/836e6fe9832c9f19146dabab29a8b55d9511459f/exporter/collector/monitoredresource.go#L201).

The `sqlserver` tests were running correctly last I checked:

```
2022/04/22 17:49:32 nonEmptySeries() iterator supplied err <nil> and series metric:{type:"workload.googleapis.com/sqlserver.user.connection.count"}  resource:{type:"gce_instance"  labels:{key:"instance_id"  value:"9188105535274022384"}  labels:{key:"project_id"  value:"stackdriver-test-143416"}  labels:{key:"zone"  value:"us-central1-b"}}  metric_kind:GAUGE  value_type:DOUBLE  points:{interval:{end_time:{seconds:1650674969  nanos:104023000}  start_time:{seconds:1650674969  nanos:104023000}}  value:{double_value:0}}
2022/04/22 17:49:32 Found representative metric, sleeping before checking remaining metrics
...
2022/04/22 17:50:42 Skipping optional or representative metric workload.googleapis.com/sqlserver.user.connection.count
2022/04/22 17:50:42 nonEmptySeries() iterator supplied err <nil> and series metric:{type:"workload.googleapis.com/sqlserver.transaction_log.flush.data.rate"  labels:{key:"sqlserver_database_name"  value:"master"}}  resource:{type:"gce_instance"  labels:{key:"instance_id"  value:"9188105535274022384"}  labels:{key:"project_id"  value:"stackdriver-test-143416"}  labels:{key:"zone"  value:"us-central1-b"}}  metric_kind:GAUGE  value_type:DOUBLE  points:{interval:{end_time:{seconds:1650675029  nanos:74760000}  start_time:{seconds:1650675029  nanos:74760000}}  value:{double_value:0}}  points:{interval:{end_time:{seconds:1650674969  nanos:103151000}  start_time:{seconds:1650674969  nanos:103151000}}  value:{double_value:0}}
```

Or are you saying that if _any_ attribute key matches that prefix, then _all_ metrics passing through that exporter, regardless of _their_ attributes, might _also_ get labels added? If that's what you're saying then I assume we can just check the agent metrics from the above test run and confirm that they don't have extraneous labels, since agent metrics pass through the same exporter:

```
{
  "timeSeries": [
    {
      "metric": {
        "labels": {
          "version": "google-cloud-ops-agent-logging/2.14.0-windows-ltsc2019"
        },
        "type": "agent.googleapis.com/agent/uptime"
      },
      "resource": {
        "type": "gce_instance",
        "labels": {
          "instance_id": "9188105535274022384",
          "zone": "us-central1-b",
          "project_id": "stackdriver-test-143416"
        }
      },
      ...
```

But you're right that the tests don't check for this currently.

File: docs/sqlserver.md
Comment (jefferbrecht) [reply docs/sqlserver.md:38]: IMO the types and units make it obvious enough that these are rates. I don't view the descriptions as contradictory but I'm not against changing it to something like `Number of lock requests per second resulting in a wait.` if we want it to be more clear or if it's more consistent with how we document other such receivers.

File: docs/sqlserver.md
Comment (jefferbrecht) [reply docs/sqlserver.md:50]: While `percent_used` is clearer IMO, I don't feel strongly about renaming this; there's precedent either way (e.g. `agent.googleapis.com/memory/percent_used` vs. `workload.googleapis.com/activemq.memory.usage`).

I'll defer to @igorpeshansky, so unless there's an objection from the other reviewers then I'm fine with changing it to `percent_used`.

**Comment by jefferbrecht** [comment]
```
2022/04/25 11:12:10 Attempt 1 of sql-std-2019-win-2019 test of sqlserver finished with err=workload.googleapis.com/sqlserver.transaction_log.shrink.count: kind: expected CUMULATIVE but got GAUGE; workload.googleapis.com/sqlserver.transaction_log.growth.count: kind: expected CUMULATIVE but got GAUGE, retryable=false
```

I just reset the descriptors for those two so they should pass if you re-run it.

**Comment by StefanKurek** [comment]
@qingling128 @jefferbrecht @igorpeshansky FYI I have created a 2nd PR for merging the functionality of this PR into the existing mssql application depending on which route we go with (we can just close the PR that no longer makes sense). https://github.com/GoogleCloudPlatform/ops-agent/pull/554

**Comment by StefanKurek** [comment]
So I believe the tests are failing because I changed two metrics to be monotonic sums bug in the OTEL receiver they are still Gauges (I have a PR in OTEL to change this as well). Can you all verify that this would cause a problem, or is there a way for this ops agent PR to move forward before the OTEL change is made?

**Comment by jefferbrecht** [comment]
> So I believe the tests are failing because I changed two metrics to be monotonic sums bug in the OTEL receiver they are still Gauges (I have a PR in OTEL to change this as well). Can you all verify that this would cause a problem, or is there a way for this ops agent PR to move forward before the OTEL change is made?

Any change between GAUGE/CUMULATIVE is breaking on the Ops Agent side, so it would cause problems if we merged this and released them as GAUGE but changed them to CUMULATIVE later -- e.g. any users who pick up a release with GAUGE would need to delete their descriptors for a new release with CUMULATIVE to work correctly.

With 2.14.0 having just been released it'll probably be at least ~1 week before the next feature release, which I would hope is enough time for the upstream OTel change to go through. Otherwise the only way forward I can think of would be to disable those metrics for now and then do a follow-up PR to re-enable them once the OTel change is merged.

**Comment by StefanKurek** [comment]
@jefferbrecht would the casttosumprocessor do the trick? I added this to the pipeline here, but either it doesn't do what I was hoping it would do, or I'm using it incorrectly.

**Comment by jefferbrecht** [comment]
> @jefferbrecht would the casttosumprocessor do the trick? I added this to the pipeline here, but either it doesn't do what I was hoping it would do, or I'm using it incorrectly.

Hmm, you're right, I just took a look through the logs and it looks like it's actually sending it as CUMULATIVE and it's only failing because the test project still has the old GAUGE descriptors:

```
Metric kind for metric workload.googleapis.com/sqlserver.transaction_log.shrink.count must be GAUGE, but is CUMULATIVE.;
Metric kind for metric workload.googleapis.com/sqlserver.transaction_log.growth.count must be GAUGE, but is CUMULATIVE.
```

I just reset the descriptors again so re-running the tests should pass this time.

Yeah, casttosumprocessor should be fine IMO, I hadn't thought of that.

File: docs/sqlserver.md
Comment (StefanKurek) [reply docs/sqlserver.md]: I went ahead and just changed this. at this point I think all open issues can be resolved

**Review by jefferbrecht — APPROVED** [review]
> @jefferbrecht looks like tests passed. I'm ok with moving the focus to https://github.com/GoogleCloudPlatform/ops-agent/pull/554 if you don't think there will be any challenge from @qingling128 on it. I think you might have to reset the descriptors over there so I can run those tests again too.
> 
> Only outstanding items are whether we should have a metric name change and whether or not some descriptions should have different wording.

Sounds good. LGTM modulo resolutions to the items you mentioned.

**Comment by StefanKurek** [comment]
> > @jefferbrecht looks like tests passed. I'm ok with moving the focus to #554 if you don't think there will be any challenge from @qingling128 on it. I think you might have to reset the descriptors over there so I can run those tests again too.
> > Only outstanding items are whether we should have a metric name change and whether or not some descriptions should have different wording.
> 
> Sounds good. LGTM modulo resolutions to the items you mentioned.

@jefferbrecht in case you didn't notice I made changes for every outstanding item. At the point the last thing remaining is whether to use this PR or to use https://github.com/GoogleCloudPlatform/ops-agent/pull/554 (which adds the new functionality to the existing mssql similar to the iis PR). If everyone agrees that is the route we should go, it would be great if everyone jumped over to that one and hopefully give approvals if all looks good 😄

**Comment by qingling128** [comment]
Catching up on this. Yeah, the decision is to extend the existing receiver as V2. See the doc `[PUBLIC] Ops Agent: new IIS/MSSQL receivers` for details. We could probably link that doc somewhere in the IIS and MSSQL scope docs.

**Comment by StefanKurek** [comment]
> Catching up on this. Yeah, the decision is to extend the existing receiver as V2. See the doc `[PUBLIC] Ops Agent: new IIS/MSSQL receivers` for details. We could probably link that doc somewhere in the IIS and MSSQL scope docs.

@qingling128 ok I will close this PR and we'll just use https://github.com/GoogleCloudPlatform/ops-agent/pull/554 from now on.

**Comment by StefanKurek** [comment]
Attempting to merge https://github.com/GoogleCloudPlatform/ops-agent/pull/554 instead

**Pull request closed without merging**
